### PR TITLE
Implement tunnel server in load balancer

### LINKED ISSUE #57: Implement tunnel server in load balancer

### PR DESCRIPTION
**Pull request by MakMuftic**
**Refactored start command to run HTTP tunnel server**

### PR Checklist

- [x] I have run linter localy
- [x] I have run unit and integration tests locally
- [x] Rebased to master branch / merged master
- [x] Updated CHANGELOG.md

### Changes

- Start HTTP tunnel server on load balancer start
- Add function for fetching public IP (pings multiple services for information, and waits for 3 confirmations)
- Add multiple flags: `tunnel-port`, `tunnel-port-range`, `public-ip`
- Expand _register response_ with tunnel url

### Issues

Closes #57

### CODE REVIEW

File: cmd/start.go [hunk 1]
```diff
@@ -117,11 +134,37 @@ func init() {
...
+		&tunnelServerPort,
+		"tunnel-port",
+		"5223",
+		"[OPTIONAL] Address on which http tunnel server is running")
+
+	startCmd.Flags().StringVar(
+		&tunnelPortRange,
+		"tunnel-port-range",
+		"20000:30000",
+		"[OPTIONAL] Range of ports on which is used to open http tunnels")
```
Comment (mpetrun5) [inline cmd/start.go]: Range of ports which is

**Review by mpetrun5 — DISMISSED** [review]

File: cmd/start.go [hunk 2]
```diff
@@ -21,9 +24,13 @@ var (
 	fee        float32
 	selection  string
 	port       int32
```
Comment (mpetrunic) [inline cmd/start.go]: httpServerPort?

File: cmd/start.go [hunk 3]
```diff
@@ -117,11 +134,37 @@ func init() {
...
 }
 
 func startCommand(_ *cobra.Command, _ []string) {
 	DisplayBanner()
+
+	var tunnelURL string
+	if publicIP == "" {
+		IP, err := ip.Get()
+		if err != nil {
+			log.Error("Unable to fetch public IP ", err)
```
Comment (mpetrunic) [inline cmd/start.go]:
```suggestion
			log.Error("Unable to fetch public IP address. Please set one explicitly!", err)
```
You could put that as fatal log

File: cmd/start.go [hunk 4]
```diff
@@ -117,11 +134,37 @@ func init() {
...
+			return
+		}
+		tunnelURL = fmt.Sprintf("%s:%s", IP.String(), tunnelServerPort)
+		log.Infof("HTTP tunnel will be started on %s with port range %s", tunnelURL, tunnelPortRange)
```
Comment (mpetrunic) [inline cmd/start.go]:
```suggestion
		log.Infof("Tunnel server will listen on %s and connect tunnels on port range %s", tunnelURL, tunnelPortRange)
```

File: internal/httptunnel/httptunnel.go [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"fmt"
+	"github.com/NodeFactoryIo/vedran/internal/auth"
+	"github.com/NodeFactoryIo/vedran/pkg/http-tunnel/server"
+	log "github.com/sirupsen/logrus"
+)
+
+func StartHttpTunnelServer(serverPort string, portRange string) {
+	logger := log.WithField("context", "http-tunnel")
```
Comment (mpetrunic) [inline internal/httptunnel/httptunnel.go]:
```suggestion
	logger := log.WithField("context", "tunnel-server")
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,16 @@
 # Changelog
 
+## [unreleased]((https://github.com/NodeFactoryIo/vedran/tree/HEAD))
+
+[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v.0.1.0...HEAD)
```
Comment (mpetrunic) [inline CHANGELOG.md]:
```suggestion
[Full Changelog](https://github.com/NodeFactoryIo/vedran/compare/v.0.1.1...HEAD)
```

File: cmd/start.go [hunk 7]
```diff
@@ -55,8 +62,12 @@ var startCmd = &cobra.Command{
...
 		}
 		// well known ports and registered ports
-		if port <= 0 && port > 49151 {
-			return errors.New("invalid port number")
+		if httpServerPort <= 0 && httpServerPort > 49151 {
+			return errors.New("invalid http server port number")
+		}
+		// valid format is PortMin:PortMax
+		if prt := strings.Split(tunnelPortRange, ":"); len(prt) != 2 {
+			return errors.New("invalid port range, should be defined as \"PortMin:PortMax\"")
```
Comment (mpetrunic) [inline cmd/start.go:72]: you should check if it's valid range also?

File: cmd/start.go [hunk 8]
```diff
@@ -100,11 +111,17 @@ func init() {
 		"[OPTIONAL] Type of selection used for choosing nodes (round-robin, random)")
 
 	startCmd.Flags().Int32Var(
-		&port,
-		"port",
+		&httpServerPort,
+		"http-server-port",
 		4000,
 		"[OPTIONAL] Port on which load balancer will be started")
```
Comment (mpetrunic) [inline cmd/start.go]:
```suggestion
		"[OPTIONAL] Port on which load balancer rpc server will be started")
```

File: cmd/start.go [hunk 9]
```diff
@@ -117,18 +134,44 @@ func init() {
 		"",
 		"[OPTIONAL] Path to logfile (default stdout)")
 
+	startCmd.Flags().StringVar(
+		&tunnelServerPort,
+		"tunnel-port",
+		"5223",
+		"[OPTIONAL] Address on which http tunnel server is running")
```
Comment (mpetrunic) [inline cmd/start.go]:
```suggestion
		"[OPTIONAL] Address on which tunnel server is listening")
```

**Review by mpetrunic — COMMENTED** [review]
Let's just call if "tunnel server" instead of "http tunnel server". If I'm not mistaken, same server can be later use to open udp tunnel?

**Review by mpetrunic — APPROVED** [review]

**Review by mpetrun5 — APPROVED** [review]

**Pull request merged**